File: src/controller/combined.controller.ts
import {
  Controller,
  Get,
  Post,
  Delete,
  Param,
  Body,
  HttpCode,
  HttpStatus,
  UsePipes,
  ValidationPipe,
} from '@nestjs/common';
import { ChoreService } from '../service/chore.service';
import { Subtask } from '../model/subtask.entity';
import { SubtaskDTO } from './dto/outgoing/subtask.dto';
import { NewSubtaskDTO } from './dto/incoming/newSubtask.dto';
import { ChoreDTO } from './dto/outgoing/chore.dto';

@Controller('chores/:choreId/subtasks')
export class SubtaskController {
  constructor(private readonly choreService: ChoreService) {}

  @Get()
  async findAll(@Param('choreId') choreId: number): Promise<SubtaskDTO[]> {
    const chore = await this.choreService.findById(choreId);
    return chore.subtasks;
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  @UsePipes(new ValidationPipe({ transform: true }))
  async create(
    @Param('choreId') choreId: number,
    @Body() subtaskData: NewSubtaskDTO
  ): Promise<ChoreDTO> {
    const chore = await this.choreService.findById(choreId);
    const subtask = { ...subtaskData, chore } as Subtask;
    chore.subtasks.push(subtask);
    await this.choreService.update(choreId, chore);
    return chore;
  }

  @Delete(':subtaskId')
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(
    @Param('choreId') choreId: number,
    @Param('subtaskId') subtaskId: number
  ): Promise<void> {
    const chore = await this.choreService.findById(choreId);
    chore.subtasks = chore.subtasks.filter(subtask => subtask.id !== subtaskId);
    await this.choreService.update(choreId, chore);
  }
}

File: src/controller/combined_controller.py
from flask import Blueprint, request, jsonify, abort
from marshmallow import ValidationError

from src.controller.dto.incoming_dto import new_subtask_schema
from src.controller.dto.outgoing_dto import subtask_schema, chore_schema
from src.service.chore_service import get_chore_by_id, add_subtask_to_chore, remove_subtask_from_chore
from src.service.exceptions import ChoreNotFoundException

combined_bp = Blueprint('subtask', __name__)


@combined_bp.route('', methods=['GET'])
def get_all(chore_id):
    try:
        chore = get_chore_by_id(chore_id)
        return jsonify(subtask_schema.dump(chore.subtasks, many=True))
    except ChoreNotFoundException:
        abort(404, 'Chore not found')


@combined_bp.route('', methods=['POST'])
def create(chore_id):
    try:
        subtask_data = new_subtask_schema.load(request.json)
        chore = add_subtask_to_chore(chore_id, subtask_data)
        return jsonify(chore_schema.dump(chore))
    except ValidationError as err:
        abort(400, err.messages)


@combined_bp.route('/<int:subtask_id>', methods=['DELETE'])
def delete(chore_id, subtask_id):
    remove_subtask_from_chore(chore_id, subtask_id)
    return '', 200
